src/Microsoft.AspNetCore.Razor/CodeGenerators/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs

```diff
@@ -93,30 +94,33 @@ protected override void Visit(TagHelperChunk chunk)
 
             if (!Context.Host.DesignTimeMode)
             {
-                PreAllocateUnboundTagHelperAttributes(chunk);
+                PreAllocateTagHelperAttributes(chunk);
             }
 
             // We need to dive deeper to ensure we pick up any nested tag helpers.
             Accept(chunk.Children);
         }
 
-        private void PreAllocateUnboundTagHelperAttributes(TagHelperChunk chunk)
+        private void PreAllocateTagHelperAttributes(TagHelperChunk chunk)
         {
             var boundAttributes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
             for (var i = 0; i < chunk.Attributes.Count; i++)
             {
                 var attribute = chunk.Attributes[i];
-                var hasAssociatedDescriptors = chunk.Descriptors.Any(descriptor =>
+                var associatedDescriptors = chunk.Descriptors.Where(descriptor =>
```

Comment:
Sorry, I failed at explaining before. I be :tired_face:. You don't utilize the full descriptors from here on down. Just do:

```C#
var associatedDescriptors = chunk.Descriptors.SelectMany(descriptor.Attributes).Where(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key));
```

You can then pass the attribute descriptors through and verify them without having to query again for name matches.